pyiron_contrib/tinybase/executor.py

```diff
@@ -0,0 +1,201 @@
+        if isinstance(state, str):
+            state = RunMachine.Code(state)
+        self._callbacks[state] = callback
+
+    def observe(self, state: Union[str, Code], callback):
+        if isinstance(state, str):
+            state = RunMachine.Code(state)
+        self._observers[state].append(callback)
+
+    def goto(self, state: Union[str, Code], **kwargs):
+        if isinstance(state, str):
+            state = RunMachine.Code(state)
+        self._state = state
+        self._data = {}
+        self._data.update(kwargs)
+        for obs in self._observers[state]:
+            obs(self._data)
+
+    def step(self, state: Union[str, Code, None] = None, **kwargs):
+        if state is not None:
+            self.goto(state, **kwargs)
+        self._callbacks.get(self._state, lambda: None)()
+
+
+class Executor:
+
+    def __init__(self, nodes):
+        self._nodes = nodes
+        self._run_machine = RunMachine("init")
+
+        # Set up basic flow
+        self._run_machine.on("init", self._run_init)
+        # exists mostly to let downstream code hook into it, e.g. to write input files and such
+        self._run_machine.on("ready", self._run_ready)
+        self._run_machine.on("running", self._run_running)
+        # exists mostly to let downstream code hook into it, e.g. to read output files and such
+        self._run_machine.on("collect", self._run_collect)
+        self._run_machine.on("finished", self._run_finished)
+
+        # keeping track of run times
+        self._running_start = None
+        self._collect_start = None
+        self._finished_start = None
+        self._run_machine.observe("running", self._time_running)
+        self._run_machine.observe("collect", self._time_collect)
+        self._run_machine.observe("finished", self._time_finished)
+
+    def _time_running(self, _data):
+        self._running_start = time.monotonic()
+
+    def _time_collect(self, _data):
+        self._collect_start = time.monotonic()
+        self._run_time = self._collect_start - self._running_start
+
+    def _time_finished(self, _data):
+        self._finished_start = time.monotonic()
+        self._collect_time = self._finished_start - self._collect_start
+
+    @property
+    def nodes(self):
+        return self._nodes
+
+    def _run_init(self):
+        if all(node.input.check_ready() for node in self.nodes):
+            self._run_machine.step("ready")
+        else:
+            logging.info("Node is not ready yet!")
+
+    def _run_ready(self):
+        self._run_machine.step("running")
+
+    def _run_running(self):
+        status, output = zip(*[node.execute() for node in self.nodes])
+        self._run_machine.step("collect", status=status, output=output)
+
+    def _run_collect(self):
+        self._run_machine.step("finished",
+                status=self.status,
+                output=self.output,
+        )
+
+    def _run_finished(self):
+        pass
+
+    def run(self):
+        self._run_machine.step()
```

Comment:
I find the run machine architecture is still not 100% transparent, but that this is already a big step up from the current schema -- nice job.

Reply:
Yeah, the relationship between the executors and the run state machine and how the state machine handles and exposes `_data` is also not quite solidified for me.  You'll see a few places still where `TinyJob` needs to reach inside all of this.